packages/infra/client/isolate-v8-runner/src/ext/kv.rs

```diff
@@ -113,37 +78,120 @@ pub fn op_rivet_kv_get_batch(
 						.map(|key| {
 							let tx = tx.clone();
 							async move {
-								Ok(tx.get(key.as_bytes(), false).await?.map(|data| (key, data)))
+								Ok(tx
+									.get(key.as_bytes(), false)
+									.await?
+									.map(|data| (key, data.to_vec().into())))
 							}
 						})
 						.buffer_unordered(16)
 						.try_filter_map(|x| std::future::ready(Ok(x)))
-						.try_collect::<Vec<_>>()
+						.try_collect::<HashMap<_, _>>()
 						.await
 				}
 			})
 			.await?;
 
-		data.into_iter()
-			.map(|(key, data)| {
-				let output = match options.format {
-					OutputFormat::Text => Output::Text(
-						std::str::from_utf8(&data)
-							.with_context(|| {
-								format!("failed to deserialize value as text for key {key:?}")
-							})?
-							.to_string(),
-					),
-					OutputFormat::Json => Output::Json(
-						serde_json::from_slice::<serde_json::Value>(&data).with_context(|| {
-							format!("failed to deserialize value as JSON for key {key:?}")
-						})?,
-					),
-					OutputFormat::ArrayBuffer => Output::Buffer(data.to_vec().into()),
-				};
-
-				Ok(output)
+		Ok(data)
+	})
+}
+
+#[op2(async)]
+pub fn op_rivet_kv_put(
+	state: &mut OpState,
+	#[string] key: String,
+	#[buffer] value: JsBuffer,
+) -> Result<impl Future<Output = Result<(), AnyError>>, AnyError> {
+	validate_key(&key)?;
+	validate_value(&key, &value)?;
+
+	let db = state.borrow::<Arc<fdb::Database>>().clone();
+
+	Ok(async move {
+		let bkey = key.as_bytes();
+
+		db.run(|tx, _maybe_committed| {
+			let value = value.clone(); // Creates a new ref, does not clone data
+
+			async move {
+				tx.set(bkey, &value);
+				Ok(())
+			}
+		})
+		.await?;
+
+		Ok(())
+	})
+}
+
+#[op2(async)]
+pub fn op_rivet_kv_put_batch(
+	state: &mut OpState,
+	#[serde] obj: HashMap<String, JsBuffer>,
+) -> Result<impl Future<Output = Result<(), AnyError>>, AnyError> {
+	for (key, value) in &obj {
+		validate_key(&key)?;
+		validate_value(&key, &value)?;
+	}
+
+	let db = state.borrow::<Arc<fdb::Database>>().clone();
+
+	Ok(async move {
+		db.run(|tx, _maybe_committed| {
+			let obj = obj.clone();
+
+			async move {
+				for (key, value) in obj {
+					tx.set(key.as_bytes(), &value);
+				}
+
+				Ok(())
+			}
+		})
+		.await?;
+
+		Ok(())
+	})
+}
+
+#[op2(async)]
+pub fn op_rivet_kv_delete(
+	state: &mut OpState,
+	#[string] key: String,
+) -> Result<impl Future<Output = Result<bool, AnyError>>, AnyError> {
+	validate_key(&key)?;
+
+	let db = state.borrow::<Arc<fdb::Database>>().clone();
+
+	Ok(async move {
+		let bkey = key.as_bytes();
+
+		let existed = db
+			.run(|tx, _maybe_committed| async move {
+				let existed = tx.get(bkey, false).await?.is_some();
```

Comment:
thinking we should get rid of the bool in the response and just run clear. goal is to map kv as close to fdb as possible with as little overhead. thoughts?